lib/iris/common/lenient.py

```diff
@@ -0,0 +1,552 @@
+            Mapping of lenient client function/method, or fully qualified sting name
+            of the function/method, to one or more lenient service
+            function/methods or fully qualified string name of function/methods.
+
+        For example::
+
+            Lenient(service1, service2, client1=service1, client2=(service1, service2))
+
+        Note that, the values of these options are thread-specific.
+
+        """
+        # The executing lenient client at runtime.
+        self.__dict__["active"] = None
+        # The global lenient services state activation switch.
+        self.__dict__["enable"] = LENIENT_ENABLE_DEFAULT
+
+        for service in args:
+            self.register_service(service)
+
+        for client, services in kwargs.items():
+            self.register_client(client, services)
+
+    def __call__(self, func):
+        """
+        Determine whether it is valid for the function/method to provide a
+        lenient service at runtime to the actively executing lenient client.
+
+        Args:
+
+        * func (callable or str):
+            A function/method or fully qualified string name of the function/method.
+
+        Returns:
+            Boolean.
+
+        """
+        result = False
+        if self.__dict__["enable"]:
+            service = qualname(func)
+            if service in self and self.__dict__[service]:
+                active = self.__dict__["active"]
+                if active is not None and active in self:
+                    services = self.__dict__[active]
+                    if isinstance(services, str) or not isinstance(
+                        services, Iterable
+                    ):
+                        services = (services,)
+                    result = service in services
+        return result
+
+    def __contains__(self, name):
+        return name in self.__dict__
+
+    def __getattr__(self, name):
+        if name not in self.__dict__:
+            cls = self.__class__.__name__
+            emsg = f"Invalid {cls!r} option, got {name!r}."
+            raise AttributeError(emsg)
+        return self.__dict__[name]
+
+    def __getitem__(self, name):
+        name = qualname(name)
+        if name not in self.__dict__:
+            cls = self.__class__.__name__
+            emsg = f"Invalid {cls!r} option, got {name!r}."
+            raise KeyError(emsg)
+        return self.__dict__[name]
+
+    def __repr__(self):
+        cls = self.__class__.__name__
+        width = len(cls) + 1
+        kwargs = [
+            "{}={!r}".format(name, self.__dict__[name])
+            for name in sorted(self.__dict__.keys())
+        ]
+        joiner = ",\n{}".format(" " * width)
+        return "{}({})".format(cls, joiner.join(kwargs))
+
+    def __setitem__(self, name, value):
+        name = qualname(name)
+        cls = self.__class__.__name__
+
+        if name not in self.__dict__:
+            emsg = f"Invalid {cls!r} option, got {name!r}."
```

Comment:
The thing is that the `__setitem__` is somewhat overloaded as it's dealing with clients, services, and options. At the time I couldn't think of a succinct error message that captured all of the possibilities, because it's not quite clear what the caller is intending to do in the first place e.g., they could be attempting to set the `enable` option but misspelt it